src/Sylius/Component/Core/OrderProcessing/StateResolver.php

```diff
@@ -61,42 +79,48 @@ public function resolvePaymentState(OrderInterface $order)
      */
     public function resolveShippingState(OrderInterface $order)
     {
-        if ($order->isBackorder()) {
-            $order->setShippingState(OrderShippingStates::BACKORDER);
-
+        if (OrderShippingStates::SHIPPED === $order->getShippingState()) {
             return;
         }
-
-        $order->setShippingState($this->getShippingState($order));
-    }
-
-    /**
-     * @param OrderInterface $order
-     *
-     * @return string
-     */
-    protected function getShippingState(OrderInterface $order)
-    {
-        $states = [];
-
-        foreach ($order->getShipments() as $shipment) {
-            $states[] = $shipment->getState();
+        /** @var Collection $shipments */
+        $shipments = $order->getShipments();
+        $shipmentCount = $shipments->count();
+        $shipmentInShippedStateAmount = $shipments->filter(
+            function (ShipmentInterface $shipment) {
+                return ShipmentInterface::STATE_SHIPPED === $shipment->getState();
+            })
+            ->count()
+        ;
+        $shipmentInReadyStateAmount = $shipments->filter(
+            function (ShipmentInterface $shipment) {
+                return ShipmentInterface::STATE_READY === $shipment->getState();
+            })
+            ->count()
+        ;
+        $shipmentInCancelledStateAmount = $shipments->filter(
+            function (ShipmentInterface $shipment) {
+                return ShipmentInterface::STATE_CANCELLED === $shipment->getState();
+            })
+            ->count()
+        ;
+
+        /** @var StateMachine $stateMachine */
+        $stateMachine = $this->stateMachineFactory->get($order, OrderShippingTransitions::GRAPH);
+
+        if ($shipmentCount === $shipmentInShippedStateAmount && OrderShippingStates::SHIPPED !== $order->getShippingState()) {
+            $stateMachine->apply(OrderShippingTransitions::TRANSITION_SHIP);
         }
 
-        $states = array_unique($states);
-
-        $acceptableStates = [
-            ShipmentInterface::STATE_READY => OrderShippingStates::READY,
-            ShipmentInterface::STATE_SHIPPED => OrderShippingStates::SHIPPED,
-            ShipmentInterface::STATE_CANCELLED => OrderShippingStates::CANCELLED,
-        ];
+        if ($shipmentCount === $shipmentInReadyStateAmount && OrderShippingStates::READY !== $order->getShippingState()) {
+            $stateMachine->apply(OrderShippingTransitions::TRANSITION_REQUEST_SHIPPING);
+        }
 
-        foreach ($acceptableStates as $shipmentState => $orderState) {
-            if ([$shipmentState] == $states) {
-                return $orderState;
-            }
+        if ($shipmentCount === $shipmentInCancelledStateAmount && OrderShippingStates::CANCELLED !== $order->getShippingState()) {
+            $stateMachine->apply(OrderShippingTransitions::TRANSITION_CANCEL);
```

Comment:
`cancel` and `request_shipping` should be moved to state machine callback (take a look at [payments](url) or contact @GSadee for details).